app/lib/tool/mirror/mirror_config.dart

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert';
+
+import 'package:json_annotation/json_annotation.dart';
+import 'package:meta/meta.dart';
+import 'package:yaml/yaml.dart' as yaml;
+
+part 'mirror_config.g.dart';
+
+/// The configuration to use when creating a local (partial) mirror of pub.dev
+@JsonSerializable(explicitToJson: true, includeIfNull: false)
+class MirrorConfig {
+  List<Package> packages;
+  List<Publisher> publishers;
+  List<User> users;
+  String defaultUser;
+
+  MirrorConfig({
+    @required this.packages,
+    @required this.publishers,
+    @required this.users,
+    this.defaultUser,
+  });
+
+  factory MirrorConfig.fromJson(Map<String, dynamic> json,
+      {bool normalize = false}) {
+    final mc = _$MirrorConfigFromJson(json);
+    if (normalize) mc.normalize();
+    return mc;
+  }
+
+  factory MirrorConfig.fromYaml(String source, {bool normalize = false}) {
+    final map = json.decode(json.encode(yaml.loadYaml(source)));
+    final mc = MirrorConfig.fromJson(map as Map<String, dynamic>);
+    if (normalize) mc.normalize();
+    return mc;
+  }
+
+  Map<String, dynamic> toJson() => _$MirrorConfigToJson(this);
+
+  void normalize() {
+    packages ??= <Package>[];
+    publishers ??= <Publisher>[];
+    users ??= <User>[];
+
+    for (final publisher in publishers) {
+      publisher.members ??= <Member>[];
+
+      // copy package definitions to [MirrorConfig.packages]
+      if (publisher.packages != null) {
+        publisher.packages.forEach((p) {
+          p.publisher = publisher.name;
+          packages.add(p);
+        });
+        publisher.packages = null;
+      }
+
+      // convert admins to members
+      if (publisher.admins != null) {
+        publisher.members.addAll(
+            publisher.admins.map((e) => Member(email: e, role: 'admin')));
+        publisher.admins = null;
+      }
+
+      // add missing users
+      publisher.members?.forEach((m) => _createUserIfNeeded(m.email));
+    }
+
+    for (final package in packages) {
+      // add missing users
+      package.uploaders?.forEach(_createUserIfNeeded);
+
+      // add missing publishers
+      if (package.publisher != null) {
+        final publisher = publishers
+            .firstWhere((p) => p.name == package.publisher, orElse: () => null);
+        if (publisher == null) {
+          publishers.add(Publisher(name: package.publisher, members: []));
```

Comment:
Hmm, how about we add the defaultUser as member in these cases?

An abandon publisher should still be possible to make, by explicitly stating `members: []´ in `publishers:`... But for the automatically created publishers I suggest default user,